README.md
# MQTTGateway for Prometheus

[![circleci](https://circleci.com/gh/inuits/mqttgateway/tree/master.svg?style=shield)][circleci]
[![go report card](https://goreportcard.com/badge/github.com/inuits/mqttgateway)][goreportcard]

A project that subscribes to MQTT queues and published prometheus metrics.

```
usage: mqttgateway [<flags>]

Flags:
  --help                        Show context-sensitive help (also try --help-long and --help-man).
  --web.listen-address=":9337"  Address on which to expose metrics and web interface.
  --web.telemetry-path="/metrics"
                                Path under which to expose metrics.
  --mqtt.broker-address="tcp://localhost:1883"
                                Address of the MQTT broker.
                                The default is taken from $MQTT_BROKER_ADDRESS if it is set.
  --mqtt.topic="prometheus/#"   MQTT topic to subscribe to.
                                The default is taken from $MQTT_TOPIC if it is set.
  --mqtt.prefix="prometheus"    MQTT topic prefix to remove when creating metrics.
                                The default is taken from $MQTT_PREFIX if it is set.
  --mqtt.username=""            MQTT username.
                                The default is taken from $MQTT_USERNAME if it is set.
  --mqtt.password=""            MQTT password.
                                The default is taken from $MQTT_PASSWORD if it is set.
  --mqtt.clientid=""            MQTT client ID.
                                The default is taken from $MQTT_CLIENT_ID if it is set.
  --log.level="info"            Only log messages with the given severity or above. Valid levels: [debug, info, warn, error, fatal]
  --log.format="logger:stderr"  Set the log target and format. Example: "logger:syslog?appname=bob&local=7" or "logger:stdout?json=true"
```

## Installation from source

Requires go

```
go get -u github.com/inuits/mqttgateway
```

## Installation from binaries

Pre-built binaries are available on the [releases
page](https://github.com/inuits/mqttgateway/releases).

## Installation with Docker

[![dockeri.co](https://dockeri.co/image/mqttgateway/mqttgateway)](https://hub.docker.com/r/mqttgateway/mqttgateway)

Use a MQTT broker which runs locally:

```
$ docker run --network="host" -p 9337:9337 mqttgateway/mqttgateway
```

Use a MQTT broker which runs elsewhere:

```
$ docker run -p 9337:9337 mqttgateway/mqttgateway --mqtt.broker-address="tcp://mosquitto.local:1883"
```


## How does it work?

mqttgateway will connect to the MQTT broker at `--mqtt.broker-address` and
listen to the topics specified by `--mqtt.topic`.

By default, it will listen to `prometheus/#`.

The format for the topics is as follow:

`prefix/LABEL1/VALUE1/LABEL2/VALUE2/NAME`

A topic `prometheus/job/ESP8266/instance/livingroom/temperature_celsius` would
be converted to a metric
`temperature_celsius{job="ESP8266",instance="livingroom"}`.

If labelnames differ for a same metric, then we invalidate existing metrics and
only keep new ones. Then we issue a warning in the logs. You should avoid it.

Two other metrics are published, for each metric:

- `mqtt_NAME_last_pushed_timestamp`, the last time NAME metric has been pushed
(unix time, in seconds)
- `mqtt_NAME_push_total`, the number of times a metric has been pushed

## Security

This project does not support authentication yet but that is planned.

## Example

```
$ mosquitto_pub -m 20.2 -t prometheus/job/ESP8266/instance/livingroom/temperature_celsius
$ curl -s http://127.0.0.1:9337/metrics|grep temperature_celsius|grep -v '#'
mqtt_temperature_celsius_last_pushed_timestamp{instance="livingroom",job="ESP8266"} 1.525185129171293e+09
mqtt_temperature_celsius_push_total{instance="livingroom",job="ESP8266"} 1
temperature_celsius{instance="livingroom",job="ESP8266"} 20.2
```

## A note about the prometheus config

If you use `job` and `instance` labels, please refer to the [pushgateway
exporter
documentation](https://github.com/prometheus/pushgateway#about-the-job-and-instance-labels).

TL;DR: you should set `honor_labels: true` in the scrape config.

[circleci]:https://circleci.com/gh/inuits/mqttgateway
[goreportcard]:https://goreportcard.com/report/github.com/inuits/mqttgateway


CHANGELOG.md
## 0.0.1 / 2020-06-12

* Initial release


logger.go
// Copyright 2020 The MQTTGateway authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	mqtt "github.com/eclipse/paho.mqtt.golang"
	"github.com/prometheus/common/log"
)

func init() {
	mqtt.ERROR = &errorLogger{}
	mqtt.CRITICAL = &errorLogger{}
	mqtt.WARN = &warnLogger{}
	mqtt.DEBUG = &debugLogger{}
}

type debugLogger struct{}

func (*debugLogger) Println(v ...interface{}) {
	log.Debug(v...)
}

func (*debugLogger) Printf(s string, v ...interface{}) {
	log.Debugf(s, v...)
}

type warnLogger struct{}

func (*warnLogger) Println(v ...interface{}) {
	log.Warn(v...)
}

func (*warnLogger) Printf(s string, v ...interface{}) {
	log.Warnf(s, v...)
}

type errorLogger struct{}

func (*errorLogger) Println(v ...interface{}) {
	log.Error(v...)
}

func (*errorLogger) Printf(s string, v ...interface{}) {
	log.Errorf(s, v...)
}


exporter.go
// Copyright 2020 The MQTTGateway authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"fmt"
	"strconv"
	"strings"
	"sync"

	mqtt "github.com/eclipse/paho.mqtt.golang"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/common/log"
)

var mutex sync.RWMutex

type mqttExporter struct {
	client         mqtt.Client
	connectDesc    *prometheus.Desc
	metrics        map[string]*prometheus.GaugeVec   // hold the metrics collected
	counterMetrics map[string]*prometheus.CounterVec // hold the metrics collected
	metricsLabels  map[string][]string               // holds the labels set for each metric to be able to invalidate them
}

func newMQTTExporter() *mqttExporter {
	// create a MQTT client
	options := mqtt.NewClientOptions()
	log.Infof("Connecting to %v", *brokerAddress)
	options.AddBroker(*brokerAddress)
	if *username != "" {
		options.SetUsername(*username)
	}
	if *password != "" {
		options.SetPassword(*password)
	}
	if *clientID != "" {
		options.SetClientID(*clientID)
	}
	m := mqtt.NewClient(options)
	if token := m.Connect(); token.Wait() && token.Error() != nil {
		log.Fatal(token.Error())
	}

	// create an exporter
	c := &mqttExporter{
		client: m,
		connectDesc: prometheus.NewDesc(
			prometheus.BuildFQName(progname, "mqtt", "connected"),
			"Is the exporter connected to mqtt broker",
			nil,
			nil),
	}

	c.metrics = make(map[string]*prometheus.GaugeVec)
	c.counterMetrics = make(map[string]*prometheus.CounterVec)
	c.metricsLabels = make(map[string][]string)

	m.Subscribe(*topic, 2, c.receiveMessage())

	return c
}

func (c *mqttExporter) Describe(ch chan<- *prometheus.Desc) {
	mutex.RLock()
	defer mutex.RUnlock()
	ch <- c.connectDesc
	for _, m := range c.counterMetrics {
		m.Describe(ch)
	}
	for _, m := range c.metrics {
		m.Describe(ch)
	}
}

func (c *mqttExporter) Collect(ch chan<- prometheus.Metric) {
	mutex.RLock()
	defer mutex.RUnlock()
	connected := 0.
	if c.client.IsConnectionOpen() {
		connected = 1.
	}
	ch <- prometheus.MustNewConstMetric(
		c.connectDesc,
		prometheus.GaugeValue,
		connected,
	)
	for _, m := range c.counterMetrics {
		m.Collect(ch)
	}
	for _, m := range c.metrics {
		m.Collect(ch)
	}
}

func (e *mqttExporter) receiveMessage() func(mqtt.Client, mqtt.Message) {
	return func(c mqtt.Client, m mqtt.Message) {
		mutex.Lock()
		defer mutex.Unlock()
		t := strings.TrimPrefix(m.Topic(), *prefix)
		t = strings.TrimPrefix(t, "/")
		parts := strings.Split(t, "/")
		if len(parts)%2 == 0 {
			log.Warnf("Invalid topic: %s: odd number of levels, ignoring", t)
			return
		}
		metric_name := parts[len(parts)-1]
		pushed_metric_name := fmt.Sprintf("mqtt_%s_last_pushed_timestamp", metric_name)
		count_metric_name := fmt.Sprintf("mqtt_%s_push_total", metric_name)
		metric_labels := parts[:len(parts)-1]
		var labels []string
		labelValues := prometheus.Labels{}
		log.Debugf("Metric name: %v", metric_name)
		for i, l := range metric_labels {
			if i%2 == 1 {
				continue
			}
			labels = append(labels, l)
			labelValues[l] = metric_labels[i+1]
		}

		invalidate := false
		if _, ok := e.metricsLabels[metric_name]; ok {
			l := e.metricsLabels[metric_name]
			if !compareLabels(l, labels) {
				log.Warnf("Label names are different: %v and %v, invalidating existing metric", l, labels)
				prometheus.Unregister(e.metrics[metric_name])
				invalidate = true
			}
		}
		e.metricsLabels[metric_name] = labels
		if _, ok := e.metrics[metric_name]; ok && !invalidate {
			log.Debugf("Metric already exists")
		} else {
			log.Debugf("Creating new metric: %s %v", metric_name, labels)
			e.metrics[metric_name] = prometheus.NewGaugeVec(
				prometheus.GaugeOpts{
					Name: metric_name,
					Help: "Metric pushed via MQTT",
				},
				labels,
			)
			e.counterMetrics[count_metric_name] = prometheus.NewCounterVec(
				prometheus.CounterOpts{
					Name: count_metric_name,
					Help: fmt.Sprintf("Number of times %s was pushed via MQTT", metric_name),
				},
				labels,
			)
			e.metrics[pushed_metric_name] = prometheus.NewGaugeVec(
				prometheus.GaugeOpts{
					Name: pushed_metric_name,
					Help: fmt.Sprintf("Last time %s was pushed via MQTT", metric_name),
				},
				labels,
			)
		}
		if s, err := strconv.ParseFloat(string(m.Payload()), 64); err == nil {
			e.metrics[metric_name].With(labelValues).Set(s)
			e.metrics[pushed_metric_name].With(labelValues).SetToCurrentTime()
			e.counterMetrics[count_metric_name].With(labelValues).Inc()
		}
	}
}

func compareLabels(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i, v := range a {
		if v != b[i] {
			return false
		}
	}
	return true
}


utilities.go
package main

import (
	"errors"
	"strings"

	pb "github.com/IHI-Energy-Storage/sparkpluggw/Sparkplug"
	mqtt "github.com/eclipse/paho.mqtt.golang"
	"github.com/golang/protobuf/proto"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/common/log"
	"github.com/prometheus/common/model"
)

// contants for various SP labels and metric names
const (
	SPNamespace  string = "sp_namespace"
	SPGroupID    string = "sp_group_id"
	SPEdgeNodeID string = "sp_edge_node_id"
	SPDeviceID   string = "sp_device_id"
	SPMQTTTopic  string = "sp_mqtt_topic"
	SPMQTTServer string = "sp_mqtt_server"
)

func sendMQTTMsg(c mqtt.Client, pbMsg *pb.Payload,
	topic string) bool {

	msg, err := proto.Marshal(pbMsg)

	if err != nil {
		log.Warnf("Failed to Marshall: %s\n", err)
		return false
	}

	token := c.Publish(topic, 0, false, msg)
	token.Wait()
	log.Debugf("%s\n", pbMsg.String())

	return true
}

func cloneLabelSet(labels prometheus.Labels) prometheus.Labels {
	newLabels := prometheus.Labels{}

	for key, value := range labels {
		newLabels[key] = value
	}

	return newLabels
}

// In order for 2 label sets to match, they have to have the exact same
// number of entries and the exact same entries orthogonal or the order
// that they are stored

func compareLabelSet(metricSet []prometheusmetric,
	newLabels []string) (bool, int) {
	returnCode := false
	returnIndex := 0
	tmpIndex := 0
	for _, existingMetric := range metricSet {

		// Make sure that both label sets have the same number of entries
		if len(existingMetric.promlabel) == len(newLabels) {

			// Initially we believe all labeles are unverified
			// As we verify we decrement, if we end up with something > 0
			// we know the set does not match

			mismatchedLabels := len(newLabels)

			for _, newLabel := range newLabels {
				// Compare the current new label to everything in existing
				// label set
				for _, existingLabel := range existingMetric.promlabel {
					if existingLabel == newLabel {
						mismatchedLabels--
						break
					}
				}
			}

			if mismatchedLabels == 0 {
				returnCode = true
				returnIndex = tmpIndex
			}
		}

		tmpIndex++
	}
	return returnCode, returnIndex
}

func createNewMetric(metricName string, metricLabels []string) *prometheus.GaugeVec {
	var newMetric prometheusmetric

	newMetric.prommetric = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: metricName,
			Help: "Metric pushed via MQTT",
		},
		metricLabels,
	)
	return newMetric.prommetric
}

func prepareLabelsAndValues(topic string) ([]string, prometheus.Labels, bool) {
	var labels []string
	t := strings.TrimPrefix(topic, *prefix)
	t = strings.TrimPrefix(t, "/")
	parts := strings.Split(t, "/")
	
	// 6.1.3 covers 9 message types, only process device data
	// Sparkplug puts 5 key namespacing elements in the topic name
	// these are being parsed and will be added as metric labels

	if (parts[2] == "DDATA") || (parts[2] == "DBIRTH") {
		if len(parts) != 5 {
			log.Debugf("Ignoring topic %s, does not comply with Sparkspec\n", t)
			return nil, nil, false
		}
	} else {
		log.Debugf("Ignoring non-device metric data: %s\n", parts[2])
		return nil, nil, false
	}

	/* See the sparkplug definition for the topic construction */
	/** Set the Prometheus labels to their corresponding topic part **/
	if labels == nil {
		labels = getLabelSet()
	}

	labelValues := prometheus.Labels{}

	// Labels are created from the topic parsing above and compared against
	// the set of labels for this metric.   If this is a unique set then it will
	// be stored and the metric will be treated as unique and new.   If the
	// metric and label set is not new, it will be updated.
	//
	// The logic for this is that the same metric name could used across
	// topics (same metric posted for different devices)

	labelValues[SPNamespace] = parts[0]
	labelValues[SPGroupID] = parts[1]
	labelValues[SPEdgeNodeID] = parts[3]
	labelValues[SPDeviceID] = parts[4]

	return labels, labelValues, true
}

func getLabelSet() []string {
	return []string{SPNamespace, SPGroupID, SPEdgeNodeID, SPDeviceID}
}

func getServiceLabelSetandValues() ([]string, map[string]string) {
	labels := []string{SPMQTTTopic, SPMQTTServer}

	labelValues := map[string]string{
		SPMQTTTopic:  *topic,
		SPMQTTServer: *brokerAddress,
	}

	return labels, labelValues
}

func getNodeLabelSetandValues(namespace string, group string,
	nodeID string) ([]string, map[string]string) {
	labels := getNodeLabelSet()
	labelValues := map[string]string{
		SPNamespace:  namespace,
		SPGroupID:    group,
		SPEdgeNodeID: nodeID,
	}

	return labels, labelValues
}

func getNodeLabelSet() []string {
	return []string{SPNamespace, SPGroupID, SPEdgeNodeID}
}
// This function acceptys MQTT metric message,
// extracts out the nested folders(if any), add those folder names in Key value labels
// and return label value sets, metrics wrt to those labelvalues and error(if any)
func getMetricName(metric *pb.Payload_Metric) ([]string, string, error) {
	var errUnexpectedType error
	var labelvalues []string

	metricName := metric.GetName()

	if strings.Contains(metricName, "/") == true && metricName != "Device Control/Rebirth" {
		parts := strings.Split(metricName, "/")
		size := len(parts)
		metricName = parts[size-1]
		for metlen := 0; metlen <= size-2; metlen++ {
			labelvalues = append(labelvalues, parts[metlen])

		}
		log.Debugf("Received message for labelvalues: %s\n", labelvalues)
	}
	metricNameL := model.LabelValue(metricName)

	if model.IsValidMetricName(metricNameL) == true {
		errUnexpectedType = nil
	} else {
		errUnexpectedType = errors.New("Non-compliant metric name")
	}

	return []string(labelvalues), string(metricNameL), errUnexpectedType
}

func convertMetricToFloat(metric *pb.Payload_Metric) (float64, error) {
	var errUnexpectedType = errors.New("Non-numeric type could not be converted to float")

	switch metric.GetDatatype() {
	case PBInt8:
		tmpLong := metric.GetIntValue()
		tmpSigned := int8(tmpLong)
		return float64(tmpSigned), nil
	case PBInt16:
		tmpLong := metric.GetIntValue()
		tmpSigned := int16(tmpLong)
		return float64(tmpSigned), nil
	case PBInt32:
		tmpLong := metric.GetIntValue()
		tmpSigned := int32(tmpLong)
		return float64(tmpSigned), nil
	case PBUInt8:
		return float64(metric.GetIntValue()), nil
	case PBUInt16:
		return float64(metric.GetIntValue()), nil
	case PBUInt32:
		return float64(metric.GetIntValue()), nil
	case PBInt64:
		// This exists because there is an unsigned consersion that
		// occurs, so moving it to an int64 allows for the sign to work properly
		tmpLong := metric.GetLongValue()
		tmpSigned := int64(tmpLong)
		return float64(tmpSigned), nil
	case PBUInt64:
		return float64(metric.GetLongValue()), nil
	case PBFloat:
		return float64(metric.GetFloatValue()), nil
	case PBDouble:
		return float64(metric.GetDoubleValue()), nil
	default:
		return float64(0), errUnexpectedType
	}
}


main.go
// Copyright 2020 The MQTTGateway authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"net/http"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/prometheus/common/log"
	"github.com/prometheus/common/version"
	"gopkg.in/alecthomas/kingpin.v2"
)

var (
	listenAddress = kingpin.Flag("web.listen-address", "Address on which to expose metrics and web interface.").Default(":9337").String()
	metricsPath   = kingpin.Flag("web.telemetry-path", "Path under which to expose metrics.").Default("/metrics").String()
	brokerAddress = kingpin.Flag("mqtt.broker-address", "Address of the MQTT broker.").Envar("MQTT_BROKER_ADDRESS").Default("tcp://localhost:1883").String()
	topic         = kingpin.Flag("mqtt.topic", "MQTT topic to subscribe to").Envar("MQTT_TOPIC").Default("prometheus/#").String()
	prefix        = kingpin.Flag("mqtt.prefix", "MQTT topic prefix to remove when creating metrics").Envar("MQTT_PREFIX").Default("prometheus").String()
	username      = kingpin.Flag("mqtt.username", "MQTT username").Envar("MQTT_USERNAME").String()
	password      = kingpin.Flag("mqtt.password", "MQTT password").Envar("MQTT_PASSWORD").String()
	clientID      = kingpin.Flag("mqtt.clientid", "MQTT client ID").Envar("MQTT_CLIENT_ID").String()
	progname      = "mqttgateway"
)

func main() {
	log.AddFlags(kingpin.CommandLine)
	kingpin.Parse()

	prometheus.MustRegister(version.NewCollector("mqttgateway"))
	prometheus.MustRegister(newMQTTExporter())

	http.Handle(*metricsPath, promhttp.Handler())
	log.Infoln("Listening on", *listenAddress)
	err := http.ListenAndServe(*listenAddress, nil)
	if err != nil {
		log.Fatal(err)
	}
}
